### ISSUE TITLE
Lint: function calls when using .expect on Result should be unwrap_or_else

### ISSUE THREAD

**Issue by alice-i-cecile** [rust-lang/rust-clippy#8820]
### What it does

Catches use of non-const function calls (like the format macro) within a `expect` method call on `Result`.

This should be replaced by `unwrap_or_else`, in order to avoid work being performed on the happy path.

### Lint Name

function_call_in_expect

### Category

perf

### Advantage

- The function called in the .expect call will be evaluated even when the method will not panic
- Using `unwrap_or_else` avoids evaluation unless the method is called

This is particularly relevant when working with allocating functions / macros, like `format!`

### Drawbacks

This is a performance micro-optimization, and makes your code slightly uglier.

Detecting whether or not a function is being used in a const way (as opposed to "if it could be const") may not be trivial. If this can't be done, then there are false positives with const functions being called.

### Example

```rust
        .expect(bevy_manifest.get_path(crate::BEVY_CREVICE))
```

Could be written as:

```rust
        .unwrap_or_else(|| bevy_manifest.get_path(crate::BEVY_CREVICE))
```
